### PR TITLE
feat: Allow multiple filters in event source mappings

### LINKED ISSUE #350: Support multiple filters in event source mappings
## Is your request related to a new offering from AWS?
Is this functionality available in the [AWS provider for Terraform](https://registry.terraform.io/providers/hashicorp/aws/latest/docs)? See [CHANGELOG.md](https://github.com/hashicorp/terraform-provider-aws/blob/main/CHANGELOG.md), too.
  - Yes ✅: please list the AWS provider version which introduced this functionality

## Is your request related to a problem? 

[Lambda supports](https://docs.aws.amazon.com/lambda/latest/dg/invocation-eventfiltering.html#filtering-syntax) up to five different filters for a single event source. It seems like the module currently only supports defining one [filter](https://github.com/terraform-aws-modules/terraform-aws-lambda/blob/8f99aaaf1af1155d71b49a377727af480af03df6/main.tf#L295).

```hcl
dynamic "filter_criteria" {
    for_each = try(each.value.filter_criteria, null) != null ? [true] : []

    content {
      filter {
        pattern = try(each.value["filter_criteria"].pattern, null)
      }
    }
  }
```
## Describe the solution you'd like.

Support multiple `filter` elements in `filter_criteria` (e.g. by providing a `list` of filters) in the `aws_lambda_event_source_mapping`:

```hcl
dynamic "filter_criteria" {
    for_each = lookup(each.value, "filter_criteria", null) != null ? [true] : []

    content {
      dynamic "filter" {
        for_each = try(each.value.filter_criteria, [])
        content {
          pattern = try(filter.value.pattern, "")
        }
      }
    }
  }
```

example usage:

```hcl
filter_criteria = [
        {
          pattern = jsonencode({
            data : {
              foo : ["bar"]
            }
          })
        },
        {
          pattern = jsonencode({
            data : {
              bar : ["foo"]
            }
          })
        }
      ]
    }
```

### PR DESCRIPTION
**Pull request by moritzzimmer**
## Description

Enhance configuration of event source mappings to support multiple filters.

## Motivation and Context

Fixes #350 

## Breaking Changes

none

## How Has This Been Tested?
- [x] I have updated at least one of the `examples/*` to demonstrate and validate my change(s)
- [x] I have tested and validated these changes using one or more of the provided `examples/*` projects

- [x] I have executed `pre-commit run -a` on my pull request

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -299,8 +299,11 @@ resource "aws_lambda_event_source_mapping" "this" {
     for_each = try(each.value.filter_criteria, null) != null ? [true] : []
 
     content {
-      filter {
-        pattern = try(each.value["filter_criteria"].pattern, null)
+      dynamic "filter" {
+        for_each = try(each.value.filter_criteria, [])
+        content {
+          pattern = try(filter.value.pattern, "")
```
Comment (antonbabenko) [inline main.tf:305]:
```suggestion
          pattern = try(filter.value.pattern, null)
```

File: main.tf [hunk 2]
```diff
@@ -299,8 +299,11 @@ resource "aws_lambda_event_source_mapping" "this" {
     for_each = try(each.value.filter_criteria, null) != null ? [true] : []
 
     content {
-      filter {
-        pattern = try(each.value["filter_criteria"].pattern, null)
+      dynamic "filter" {
+        for_each = try(each.value.filter_criteria, [])
```
Comment (antonbabenko) [inline main.tf:303]: Please add newline after `for_each`.

Comment (antonbabenko) [reply main.tf:303]: There should be a way to make it without introducing a break for existing users.

Please look at this PR where I did a similar change and added support for both types `map` and `list(map)` - https://github.com/terraform-aws-modules/terraform-aws-cloudfront/pull/91

**Review by antonbabenko — CHANGES_REQUESTED** [review]

File: main.tf (hunk 2 above)
Comment (moritzzimmer) [reply main.tf]: Tried to do this supporting both the object and list of object configuration. 

Should we use just the new notation in the examples? I kept one using the old notation for testing

File: main.tf (hunk 1 above)
Comment (moritzzimmer) [reply main.tf]: done

**Comment by moritzzimmer** [comment]
@antonbabenko I pushed changes to address your PR feedback. What do you think?

**Comment by moritzzimmer** [comment]
Ping

**Comment by moritzzimmer** [comment]
@antonbabenko @bryantbiggs any chance we get this done?

**Comment by antonbabenko** [comment]
This PR is included in version 4.10.0 :tada:

**Pull request merged**